### PR TITLE
New commandline parser

### PR DESCRIPTION
**Pull request by DatCorno**
Took what existed in Kleio and imported it here so that we can remove the parsing from the `SpaceBuilder` class and refactor it to be more extendable

### CODE REVIEW

File: src/orion/core/io/cmdline_parser.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+        self.arguments = OrderedDict()
+        self._parse_arguments(commandline)
+
+        for key, value in self.arguments.items():
+            # TODO: Support passing the same commandline but slightly different
+            # Handle positional arguments
+            if key.startswith("_"):
+                if self._already_parsed:
+                    raise RuntimeError("Cannot branch using positional arguments.")
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: This is specific to Kleio

File: src/orion/core/io/cmdline_parser.py [hunk 2]
```diff
@@ -20,28 +20,39 @@
 
 
 class CmdlineParser(object):
+    """CmdlineParser is aimed at providing a simple tool to interpret commandline arguments
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: There should be a short summary followed by the description. 
Numpy docstring-guide:
https://docs.scipy.org/doc/numpy/docs/howto_document.html#numpydoc-docstring-guide

Most specifically:
https://docs.scipy.org/doc/numpy/docs/howto_document.html#sections

File: src/orion/core/io/cmdline_parser.py [hunk 3]
```diff
@@ -20,28 +20,39 @@
...
+    for the purposes of Orion. It can transform a list of string representing arguments to their
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: It returns a list or a dictionary? There should be notes about specific cases like boolean arguments and array arguments. Also there should be a mention of the limitation of the parser for commandlines where there is optional arguments before a command, like it is for Oríon. For instance ```somecommand --optional argument subcommand --another argument``` would be incorrectly parsed as 
```python
{'optional': ['argument', 'subcommand'],
 'another': 'argument'}
```

File: src/orion/core/io/cmdline_parser.py [hunk 4]
```diff
@@ -20,28 +20,39 @@
...
     def __init__(self):
+        """Initialize the OrderedDict of arguments and the template list"""
         self.arguments = OrderedDict()
         self._already_parsed = False
         self.template = []
 
     def format(self, configuration):
+        """Recreate the string of argument using the template made at parse-time.
+        The `configuration` argument must be a `dict` compatible with the template.
+        """
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: Parameters shoud be document as described [here](https://docs.scipy.org/doc/numpy/docs/howto_document.html#sections)

Comment (bouthilx) [reply src/orion/core/io/cmdline_parser.py]: Also, same comment for one-line summary followed by description.

File: src/orion/core/io/cmdline_parser.py [hunk 5]
```diff
@@ -20,28 +20,39 @@
...
         return " ".join(self.template).format(**configuration)
 
     def parse(self, commandline):
+        """Parse the `commandline` argument to create an OrderedDict where the keys are the
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: Same comment for one-line summary followed by description.

Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: Shoud add note for the different type of arguments; positional, array, boolean, etc.

File: src/orion/core/io/cmdline_parser.py [hunk 6]
```diff
@@ -62,14 +73,15 @@ def parse(self, commandline):
...
                     template = "{" + key + "[" + str(pos) + "]}"
                     self.template.append(template)
 
         self._already_parsed = True
 
         return self.arguments
 
+    # pylint: disable=no-self-use
     def _key_to_arg(self, key):
         arg = key.replace("!!", "_").replace("??", "-")
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: Those magic values `"!!"` for `"_"`  and `"??"` for `"-" should be made attributes so that they are customizable.

File: tests/unittests/core/io/test_cmdline_parser.py [hunk 7]
```diff
@@ -85,16 +91,8 @@ def test_parse_arguments(basic_config):
     assert cmdline_parser.arguments == basic_config
 
 
-def test_parse_arguments_configuration(basic_config):
```
Comment (bouthilx) [inline tests/unittests/core/io/test_cmdline_parser.py]: Why did you remove this one?

File: tests/unittests/core/io/test_cmdline_parser.py [hunk 8]
```diff
@@ -120,6 +119,7 @@ def test_parse_arguments_bad_command():
 
 
 def test_parse_branching_arguments_template():
+    """Test the branching of args"""
```
Comment (bouthilx) [inline tests/unittests/core/io/test_cmdline_parser.py]: This we should not support, it is kleio specific. You could drop this unit-test.

File: tests/unittests/core/io/test_cmdline_parser.py [hunk 9]
```diff
@@ -144,6 +144,7 @@ def test_parse_branching_arguments_template():
 
 
 def test_parse_branching_arguments_format(monkeypatch):
+    """The the template from the branching"""
```
Comment (bouthilx) [inline tests/unittests/core/io/test_cmdline_parser.py]: Test? ;)

File: src/orion/core/io/cmdline_parser.py [hunk 10]
```diff
@@ -20,29 +20,102 @@
...
-    def __init__(self):
-        """Initialize the OrderedDict of arguments and the template list"""
+    def __init__(self, parsing_tokens=None):
+        """See `CmdlineParser` description"""
         self.arguments = OrderedDict()
         self._already_parsed = False
         self.template = []
 
+        if parsing_tokens is None:
+            parsing_tokens = {'-': '??', '_': '!!'}
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: It would be better to use update so that user can pass incomplete dicts. 

```
self.parsing_tokens = {'-': '??', '_': '!!'}
if parsing_tokens it not None:
    self.parsing_tokens.update(parsing_tokens)
```

File: src/orion/core/io/cmdline_parser.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        ----------
+        configuration : dict
+            Dictionary storing the keys and values to be passed to the `format` function.
+            This would typically be `parser.arguments` where `parser` is a `CmdlineParser`
+            instance.
+
+        Returns
+        -------
+        str
+            A recreated string of the commandline passed to Orion.
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: Of a commandline in general, not specific to Oríon.

Comment (bouthilx) [reply src/orion/core/io/cmdline_parser.py]: What about
```
The string of the original commandline parsed by the CmdlineParser, but formatted with given `configuration`.
```

Also, if you could add an example taking inspiration from the unit-tests that would be nice.

Comment (DatCorno) [reply src/orion/core/io/cmdline_parser.py]: Will do.

File: src/orion/core/io/cmdline_parser.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+        Commandline arguments' name and value(s).
+    template : list
+        List of template-ready strings (in the form 'something_{0}') for formatting.
+
+    """
+
+    def __init__(self, parsing_tokens=None):
+        """See `CmdlineParser` description"""
+        self.arguments = OrderedDict()
+        self._already_parsed = False
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: How do we handle parsing twice with the same `CmdlineParser` now? Seems like `already_parsed` is not use anywhere anymore.

File: src/orion/core/io/cmdline_parser.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+        `{'arg1': ['value1', 'value2'], 'arg2': ['value3', 'value4']}`.
+
+        File paths are extended to their absolute forms.
+
+        If the commandline contains optional arguments before a subcommand, the arguments
+        will be wrongly parsed : `somecommand --optional argument subcommand --another argument`
+        will be parsed as `{'optional': ['argument', 'subcommand'], 'another': 'argument'}`
+
+        When dealing with an empty list (of the form `--args`), the value is a boolean.
+        This indicates to the parsing process to skip this particular argument.
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: I would drop the last sentence, it only makes it confusing for the user in my opinion.

File: src/orion/core/io/cmdline_parser.py [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+        in the list. For example, `val1 val2` would be parsed as
+        `{'_pos_0': 'val1', '_pos_1': 'val2}`.
+
+        Arrays are parsed starting at the first named argument until the next one. For example :
+        `--arg1 value1 value2 --arg2 value3 value4` will be parsed as :
+        `{'arg1': ['value1', 'value2'], 'arg2': ['value3', 'value4']}`.
+
+        File paths are extended to their absolute forms.
+
+        If the commandline contains optional arguments before a subcommand, the arguments
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: Should turn this into a [warning section](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html#warnings)

File: src/orion/core/io/cmdline_parser.py (hunk 11 above)
Comment (bouthilx) [reply src/orion/core/io/cmdline_parser.py]: Bump

File: src/orion/core/io/cmdline_parser.py [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+            Dictionary holding the values of every argument. The keys are the arguments' name.
+
+        Raises
+        ------
+        ValueError
+            If there is a duplicate argument
+
+        Notes
+        -----
+        By default, all values are `str` unless their types have been changed in the meantime.
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: This is not clear. What means unless their types have been changed in the meantime?

File: src/orion/core/io/cmdline_parser.py [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+
+        """
+        return " ".join(self.template).format(**configuration)
+
+    def parse(self, commandline):
+        """Parse the `commandline` argument.
+
+        Create an OrderedDict where the keys are the names of the arguments and the values
+        are the actual values of the each argument. The arguments can be a single value or a
+        list of values. This also supports positional arguments.
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: It also creates a template for the `format()` method.

File: tests/unittests/core/io/test_cmdline_parser.py (hunk 7 above)
Comment (bouthilx) [reply tests/unittests/core/io/test_cmdline_parser.py]: Bump

File: tests/unittests/core/io/test_cmdline_parser.py (hunk 9 above)
Comment (bouthilx) [reply tests/unittests/core/io/test_cmdline_parser.py]: Oh, this we should also not support, it is also kleio specific.

File: src/orion/core/io/cmdline_parser.py [hunk 17]
```diff
@@ -0,0 +1,229 @@
+# -*- coding: utf-8 -*-
+"""
+:mod:`orion.core.io.cmdline_parser` -- Parse command line arguments for Orion
+=============================================================================
+
+.. module:: cmdline_parser
+   :platform: Unix
+   :synopsis: Parsing and building of the command line input for using script
+
+Simplify the parsing of a command line by storing every values inside an `OrderedDict`
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: Nitpicking; I think "Simplify" is irrelevant here.

File: src/orion/core/io/cmdline_parser.py [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+
+                for pos in range(len(value)):
+                    template = "{" + key + "[" + str(pos) + "]}"
+                    self.template.append(template)
+
+        self._already_parsed = True
+
+        return self.arguments
+
+    # pylint: disable=no-self-use
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: It uses self now with `_*_token`, you can remove the pylint exception.

File: src/orion/core/io/cmdline_parser.py [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+
+                if len(arg_parts) > 1 and "=".join(arg_parts[1:]).strip(" "):
+                    self.arguments[argument_name].append("=".join(arg_parts[1:]))
+
+            # If the argument did not start with `-` but we have an argument name
+            # That means that this value belongs to that argument name list
+            elif argument_name is not None and arg.strip(" "):
+                self.arguments[argument_name].append(arg)
+
+            # No argument name means we have reached them yet, so we're still in the
```
Comment (bouthilx) [inline src/orion/core/io/cmdline_parser.py]: we **didn't reach** them

File: src/orion/core/io/orion_cmdline_parser.py [hunk 20]
```diff
@@ -0,0 +1,228 @@
+# -*- coding: utf-8 -*-
+"""
+:mod:`orion.core.io.orion_cmdline_parser` -- Parse command line arguments for Orion
+===================================================================================
+
+.. module:: orion_cmdline_parser
+   :platform: Unix
+   :synopsis: Parsing and building of the command line input for using script
+
+Simplify the parsing of a command line by storing every values inside an `OrderedDict`
```
Comment (bouthilx) [inline src/orion/core/io/orion_cmdline_parser.py]: This should be rewritten for the specific case of OrionCmdlineParser

File: src/orion/core/io/orion_cmdline_parser.py [hunk 21]
```diff
@@ -0,0 +1,228 @@
...
+        self.priors_only = OrderedDict()
+        self.file_config = OrderedDict()
+        self.augmented_config = OrderedDict()
+
+        self.config_prefix = config_prefix
+        self.file_config_path = None
+        self.converter = None
+
+        self._extraction_method = {dict: self._extract_dict,
+                                   list: self._extract_list,
```
Comment (bouthilx) [inline src/orion/core/io/orion_cmdline_parser.py]: We could miss tuples

File: src/orion/core/worker/experiment.py [hunk 22]
```diff
@@ -114,6 +115,7 @@ def __init__(self, name):
         self.pool_size = None
         self.max_trials = None
         self.algorithms = None
+        self.parser = OrionCmdlineParser()
```
Comment (bouthilx) [inline src/orion/core/worker/experiment.py]: I guess the Consumer should be modified to now use this parser.

File: tox.ini [hunk 23]
```diff
@@ -81,7 +81,7 @@ deps =
 usedevelop = True
 commands =
     pip install -U {toxinidir}/tests/functional/gradient_descent_algo
-    python setup.py test --addopts '-vvv --exitfirst --capture=no --looponfail {posargs}'
```
Comment (bouthilx) [inline tox.ini]: Why did you drop the --capture=no?

File: src/orion/core/io/cmdline_parser.py (hunk 11 above)
Comment (DatCorno) [reply src/orion/core/io/cmdline_parser.py]: This change is inside unpushed commits.

File: src/orion/core/io/cmdline_parser.py (hunk 13 above)
Comment (DatCorno) [reply src/orion/core/io/cmdline_parser.py]: Will do.

File: src/orion/core/io/cmdline_parser.py (hunk 14 above)
Comment (DatCorno) [reply src/orion/core/io/cmdline_parser.py]: Noted.

File: src/orion/core/io/cmdline_parser.py (hunk 16 above)
Comment (DatCorno) [reply src/orion/core/io/cmdline_parser.py]: Done.

File: src/orion/core/io/cmdline_parser.py (hunk 17 above)
Comment (DatCorno) [reply src/orion/core/io/cmdline_parser.py]: Would "Handles" be better here? There is also a "simple" inside the class description.

File: src/orion/core/io/cmdline_parser.py (hunk 19 above)
Comment (DatCorno) [reply src/orion/core/io/cmdline_parser.py]: Fixed.

File: src/orion/core/io/cmdline_parser.py (hunk 18 above)
Comment (DatCorno) [reply src/orion/core/io/cmdline_parser.py]: Removed.

File: src/orion/core/io/orion_cmdline_parser.py [hunk 24]
```diff
@@ -0,0 +1,228 @@
...
+        list
+            A list composed of the same elements as `args` augmented with the new form
+            of the priors.
+        """
+        replaced = []
+        for item in args:
+            if item.startswith('-'):
+                parts = item.split('~')
+
+                if parts[0].startswith('--') and len(parts[0]) == 3:
```
[the remark this answers is not in the document]
Comment (DatCorno) [reply src/orion/core/io/orion_cmdline_parser.py]: Yes, this would be split as [`--x`, `prior(yyy)`] which is not valid anymore.

File: src/orion/core/io/orion_cmdline_parser.py [hunk 25]
```diff
@@ -0,0 +1,228 @@
...
+                    replaced.append('orion~' + parts[1])
+            else:
+                replaced.append(item)
+
+        return replaced
+
+    def _build_priors_only(self, configuration):
+        """Create OrderedDict from priors only.
+
+        Loop through every commandline arguments and check if it might correspond the a prior.
```
[the remark this answers is not in the document]
Comment (DatCorno) [reply src/orion/core/io/orion_cmdline_parser.py]: Fixed.

File: tox.ini (hunk 23 above)
Comment (DatCorno) [reply tox.ini]: If you recall, the `capture=no` led to a bug we had some months ago.

**Comment by bouthilx** [comment]
A new PR should be done following #155
